specification/draft-2/README.md

```diff
@@ -0,0 +1,687 @@
+The `source` property can also be an array of port identifiers, if one wishes to connect multiple
+source ports to same destination. In this case, CWL implementations will create an array from all
+incoming values. Example:
+
+```yaml
+steps:
+  # ...
+  - id: step2
+    inputs:
+      - id: step2/in_files
+        connect:
+          source: [step0/out_file, step1/out_file]
+```
+
+
+
+# Feature Extensions
+
+CWL specifies additional features to support describing more complex tools and workflows.
+Each of these features, if used, needs to be listed explicitly in the `requirements` section.
+
+
+## Custom types
+
+By requiring a `CustomTypesRequirement`, you can use the `schemaDefs` property
+to enumerate named Avro types.
+
+For example, a tool can define an input of complex type like this:
+
+ ```yaml
+ class: CommandLineTool
+ requirements:
+   - class: CustomTypesRequirement
+ schemaDefs:
+   - name: NamedGroupOfFiles
+     type: record
+     fields:
+       - name: groupName
+         type: string
+       - name: fileList
+         type:
+           type: array
+           items: File
+inputs:
+  - id: fileGroups
+    type:
+      type: array
+      items: NamedGroupOfFiles
+# ...
+ ```
+
+and the Job object might look like this:
+
+```yaml
+inputs:
+  fileGroups:
+    - groupName: group1
+      fileList: [] # File objects listed here
+    - groupName: group2
+      fileList: [] # File objects listed here
+```
+
+
+## File metadata
+
+By requiring `FileMetadataRequirement`, you can annotate produced files with key-value pairs
+and attach "secondary" files that follow a naming convention to them (usually indices).
+
+For example, a tool that creates an index for a BAM file can describe its output like this:
+
+```yaml
+class: CommandLineTool
+label: bamtools index
+requirements:
+  - class: FileMetadataRequirement
+  - class: ExpressionsRequirement
+inputs:
+  - id: bamFile
+    type: File
+    binding:
+      position: 1
+outputs:
+  - id: indexed
+    type: File
+    binding:
+      glob:
+        class: Expression
+        script: $job.inputs.bamFile.path
+      fileMetadata:
+        cwlm:format: BAM
+      secondaryFiles: [.bai]
+baseCmd: [bamtools, index]
+```
+
+The "bamtools index" command will create a .bai file in the same directory where BAM file is located.
+The tool is described as taking a BAM file and producing an "indexed" BAM file, which is the same file
+but with a .bai file "attached" to it. It also sets the metadata key `cwlm:format` to "BAM".
+
+Metadata keys can be any strings, but we encourage use of URIs, specifically those in the cwlm namespace
+(which is still not defined - work in progress).
+
+The `secondaryFiles` property is a list of file extensions in form of `.ext` that get appended to
+path of master file. This tells the CWL implementation to always make those files available in the
+same directory as the master file. To replace the extension (rather than append) use `^.ext` -
+e.g. `^.dict` for FASTA files.
```

Comment:
You are right, for the last case it would better to accomplish it through something like a RegexExpressionEngine, and allow "secondaryFiles" to be generated from arbitrary expressions if necessary.